x_twist_compensate: Add X twist compensation module

koonweee opened the pull request:
# Context
Printers with a probe that have a probe y_offset may give inaccurate probe results due to a twist in the probe's X axis. This is common in printers with designs like the Prusa MK3, Sovol 06 etc, as described under [probe location bias](https://www.klipper3d.org/Probe_Calibrate.html#location-bias-check) in the Klipper documentation. It may result in probe operations such as Bed Mesh, Screws Tilt Adjust, Z Tilt Adjust etc returning inaccurate representations of the bed.

The following diagram (by @MauroGil66) and video help to visualize the issue
![image](https://user-images.githubusercontent.com/34903259/218170885-9d42991a-9a6e-464a-ba86-b0ee32745570.png)
[Video "Visualization of x-sled-rotation", by Elhanan Maayan](https://www.youtube.com/watch?v=rPU4ZAPJTNc)

This issue is also often [reported on Marlin](https://github.com/MarlinFirmware/Marlin/issues/22791), and some of the users on the Klipper discourse [[Post 1](https://klipper.discourse.group/t/qgl-with-bed-mesh-still-cant-get-gantry-level-to-bed/4584/9), [Post 2](https://klipper.discourse.group/t/impossible-bed-mesh-leveling-cr10spro-v2/204/75)]

# Fix
The fix implemented adapts math from the [fix implemented in Marlin](https://github.com/MarlinFirmware/Marlin/pull/23238). The user can manually calibrate the z-offset at a few points along the x axis, and linear interpolation is done to figure out the appropriate z-offset at all other points.

`probe.py` is modified such that the z-compensation is applied for all calls to the probe to determine a z-height. This allows modules that use the probe such as bed_mesh and z_tilt_adjust to benefit from the calibrated probe z-heights.

# Testing
Admittedly, real world testing is currently limited on this. I have personally tested it to work wonderfully on my own SV06, and have posted it in a few Discord channels/the Klipper discourse for more users to test it out.

All comments, criticism, and feedback are very welcome! Thank you.

Signed-off by: Jeremy Tan jeremytkw98@gmail.com

koonweee commented:
Installation instructions for interested testers:
```
# change directory to your klipper install folder on your RPI/other klipper host
cd klipper

# add my klipper fork as a remote, and fetch it
git remote add koonweee https://github.com/koonweee/klipper.git
git fetch koonweee

# checkout the new "x_twist_compensation" module and the modified "probe" module
git checkout remotes/koonweee/jt/x-twist-compensation-simple klippy/extras/x_twist_compensation.py klippy/extras/probe.py
```

heqstf commented:
Hi @koonweee,

 Is this X twist compensation module,rearranged to XY? i.e: cross xy could have a twisted Y too. So the given reads could be a misrepresentation of your xy plane twist.
 
 Cheers.

LHARD commented:
Hi @koonweee ,

I just tested your work.

I did everything as described, but for me it didn't work; at least not without a small modification:
I had to invert the values in the printer.cfg to make the mesh look similar to my manual mesh.
Didn't do a new test-print since then. But first test was just as off as the mesh looked like in mainsail.

I use a Geeetech A30M with a creality sprite pro extruder (and obviously klipper).

koonweee commented:
> Hi @koonweee,
> 
> Is this X twist compensation module, be rearranged to XY? i.e: cross xy could have a twisted Y too. So the given reads could be a misrepresentation of your xy plane twist.
> 
> Cheers.

Hey! It is possible to generalize the module for the y axis too. The compensation required would be a summation of both the x and y axis if both are twisted. I believe X twist issues are more common however

koonweee commented:
> 

Hey thanks for testing the module. Could you elaborate on your probe setup on your printer? (eg. what probe is used)

koonweee commented:
> Hi @koonweee ,
> 
> I just tested your work.
> 
> I did everything as described, but for me it didn't work; at least not without a small modification: I had to invert the values in the printer.cfg to make the mesh look similar to my manual mesh. Didn't do a new test-print since then. But first test was just as off as the mesh looked like in mainsail.
> 
> I use a Geeetech A30M with a creality sprite pro extruder (and obviously klipper).

Hey thanks for testing the module. Could you elaborate on your probe setup on your printer? (eg. what probe is used)

fauks commented:
I will be testing this solution this week/weekend. I have mesh issues that I have narrowed down to X extrusion twisting. Unfortunately I cannot manually level this bed so I'm hoping this will work and I will report back my findings and printer information.

LHARD commented:
Hey, I edited my previous post with the missing info.

Sent with Spark
Am 12. Feb. 2023, 13:45 +0100 schrieb Jeremy Tan ***@***.***>:
> > Hi @koonweee ,
> > I just tested your work.
> > I did everything as described, but for me it didn't work; at least not without a small modification: I had to invert the values in the printer.cfg to make the mesh look similar to my manual mesh. Didn't do a new test-print since then. But first test was just as off as the mesh looked like in mainsail.
> > I use a Geeetech A30M with a creality sprite pro extruder (and obviously klipper).
> Hey thanks for testing the module. Could you elaborate on your probe setup on your printer? (eg. what probe is used)
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>

koonweee commented:
> Hi @koonweee ,
> 
> I just tested your work.
> 
> I did everything as described, but for me it didn't work; at least not without a small modification: I had to invert the values in the printer.cfg to make the mesh look similar to my manual mesh. Didn't do a new test-print since then. But first test was just as off as the mesh looked like in mainsail.
> 
> I use a Geeetech A30M with a creality sprite pro extruder (and obviously klipper). My Probe is a Geeetech 3D Touch V3.2 Pro, mounted at the stock position for the CRTouch. It is quite far away from the nozzle: x_offset: -33 y_offset: -46

Just a guess on my part, but does `PROBE_CALIBRATE` work correctly on your setup? eg. does it probe a point, and then corrctly move the nozzle over that point for manual z offset calibration?

koonweee commented:
> I will be testing this solution this week/weekend. I have mesh issues that I have narrowed down to X extrusion twisting. Unfortunately I cannot manually level this bed so I'm hoping this will work and I will report back my findings and printer information.

Thank you and please let me know! out of curioisty, why cant your bed be manually levelled?

fauks commented:
>>
> Thank you and please let me know! out of curiosity, why cant your bed be manually levelled?

The build plate is hard mounted. I can actually level it since I have modified the bed, but the screws are located under the flex plate which makes tramming via paper/feeler gauge difficult as I would have to move the nozzle every time I needed to make an adjustment. There are also 12 mounting points instead of your typical 4. I however can use screws_tilt_calculate but I think this was not working right because of possible twist in the X axis. 

I tried it last night, it was more simple than I thought it would be. I had to manually adjust the values myself across 3 points but it did indeed work. I didn't have any non-magnetic feeler gauges so I first did the paper test and got better results than what I was getting prior, and then manually adjusted the values myself until now the whole plate has a great first layer. 

What method would you use for a bed that cannot be trammed? Like, what order of leveling with only access to screws_tilt_adjust. Should I run that without x_twist_compensation enabled first, then preform x_twist after, and then re-level using screws_tilt_calculate? 

I'm thinking the order I did things may have made the results not perfect first try around, or maybe its because I was using the paper test which can be pretty inaccurate. 

Anyways, thank you so much for this! I haven't had a good first layer in almost a year and have been ready to trash this printer so I'm pretty stoked.

koonweee commented:
> > > 
> > 
> > 
> > Thank you and please let me know! out of curiosity, why cant your bed be manually levelled?
> 
> The build plate was hard mounted. I have modified the bed so I can manually adjust the bed, but the screws are located under the flex plate which makes tramming via paper/feeler gauge difficult as I would have to move the nozzle and flex plate every time I needed to make an adjustment. The underside "nut" is just tapped into the Y carriage so I can't use a normal wheel nut to adjust from below. There are also 12 mounting points instead of your typical 4. I however can use screws_tilt_calculate since I can make all 12 adjustments, then re-run it until "level" but I think this was not working right because of possible twist in the X axis. I was having issues where the far right was nearly touching the bed, and the far left was stringing. My BLTouch sensor offset is x: -41, y: +25 from the nozzle. So my probe is actually behind the nozzle which explains why I'm see the opposite pattern as people with a probe in front of the nozzle.
> 
> I tried it last night, it was more simple than I thought it would be. I had to manually adjust the values myself across 3 points but it did indeed work. I didn't have any non-magnetic feeler gauges so I first did the paper test and got better results than what I was getting prior, and then manually adjusted the values myself until now the whole plate has a great first layer.
> 
> What method would you use for a bed that cannot be trammed? Like, what order of leveling with only access to screws_tilt_adjust. Should I run that without x_twist_compensation enabled first, then preform x_twist after, and then re-level using screws_tilt_calculate?
> 
> I'm thinking the order I did things may have made the results not perfect first try around, or maybe its because I was using the paper test which can be pretty inaccurate.
> 
> Anyways, thank you so much for this! I haven't had a good first layer in almost a year and have been ready to trash this printer so I'm pretty stoked.
> 
> EDIT: I also used the value recommended for probe_offset which worked nearly perfect. I adjusted it a bit after but no more than 0.02mm. I have not tried to use probe_calibrate after doing any of the above. I saw you asked another user about that.

Great to hear! I actually have a similar setup with my sv06 where the 4 mounting points have been hardmounted with nylock nuts but I can use bed_screws_calculate to figure out how much to adjust them. As you noted, using this compensation allows you to use bed_screws_calculate, since the probe readings are now correct. 

Since the x_twist_compensation profile corrects your probe readings, I recommend having compensation profile loaded before you perform any bed_screws_calculate or bed_mesh.

I think the slight inaccuracy you saw was indeed from using the paper test. I also use the paper since I do not have any feeler gauges and also have to tweak the recommended probe_offset and generated compensation values ever so slightly to get a perfect first layer. With a feeler gauge, I'm quite sure the generated compensation would be even more accurate on the first try.

Do not worry about the probe_calibrate, I asked the other user to perform it just to make sure that their probe setup is correct since they were facing unexpected results (the x twist correction generated was inverted from what was expected)

fauks commented:
>>
> Great to hear! I actually have a similar setup with my sv06 where the 4 mounting points have been hardmounted with nylock nuts but I can use bed_screws_calculate to figure out how much to adjust them. As you noted, using this compensation allows you to use bed_screws_calculate, since the probe readings are now correct.
> 
> Since the x_twist_compensation profile corrects your probe readings, I recommend having compensation profile loaded before you perform any bed_screws_calculate or bed_mesh.
> 
> I think the slight inaccuracy you saw was indeed from using the paper test. I also use the paper since I do not have any feeler gauges and also have to tweak the recommended probe_offset and generated compensation values ever so slightly to get a perfect first layer. With a feeler gauge, I'm quite sure the generated compensation would be even more accurate on the first try.
> 
> Do not worry about the probe_calibrate, I asked the other user to perform it just to make sure that their probe setup is correct since they were facing unexpected results (the x twist correction generated was inverted from what was expected)

I had been meaning to pick up a set of plastic feeler gauges or at least some non magnetic. Maybe I have more of an excuse for it now!

I'm about to start a full plate of prints which I haven't been able to do in a very long time :)

LHARD commented:
> > Hi @koonweee ,
> > I just tested your work.
> > I did everything as described, but for me it didn't work; at least not without a small modification: I had to invert the values in the printer.cfg to make the mesh look similar to my manual mesh. Didn't do a new test-print since then. But first test was just as off as the mesh looked like in mainsail.
> > I use a Geeetech A30M with a creality sprite pro extruder (and obviously klipper). My Probe is a Geeetech 3D Touch V3.2 Pro, mounted at the stock position for the CRTouch. It is quite far away from the nozzle: x_offset: -33 y_offset: -46
> 
> Just a guess on my part, but does `PROBE_CALIBRATE` work correctly on your setup? eg. does it probe a point, and then corrctly move the nozzle over that point for manual z offset calibration?
> 
> edit: Let me know about the above, but I will also write some test cases soon which should catch any anomalies in the code

PROBE_CALIBRATE works normal incl. moving the nozzle.
Just some random info: my z-endstop is some mm **above** my z=0 point; so after homing the nozzle will go down

murdock62 commented:
Hi @koonweee,

I performed several tests. Since the result was not yet perfect and my mesh is done on 9 points, 
I performed an X_TWIST_CALIBRATE N_POINTS= 9.

The calibration went well, but when I want to perform a BED_MESH_CALIBRATE it probes the first point 
well and when it goes to the second klipper crashes and sends me this message :

"list indices must be integers, not float"


By leaving default (so with 3 points) no worries. But the result is not perfect. 
(with a 350mm bed this is probably normal)

Thank you for your help.

murdock62 commented:
Hi @koonweee,

I looked quickly.
I think by simply replacing:

`interpolate_i = BedMesh.constrain(interpolate_i, 0, n_points - 2)`
by
`interpolate_i = int(BedMesh.constrain(interpolate_i, 0, n_points - 2))`

Must do the job. But I'm not a python expert.

murdock62 commented:
I created a small simulation program based on your function (with the addition of the int conversion of the index) 
when calling the "_probe" function of klipper and it looks good, it gives the following results:

![Simple_X_Twist correction](https://user-images.githubusercontent.com/11955934/220286336-a3ac0c04-3ef7-4ffd-8f5f-83307fdd7ea7.jpg)

The way "interpolate_t" is calculated is used in the calculation of the linear correction gives a small error but 
that is really not far we are talking in the worst case of an error of about 3 thousandths of mm.

Now all that remains is to test with 9 points in real on my machine.

Many thanks for your work.

koonweee commented:
> 

Hey! Did you get round to testing it on your machine?

murdock62 commented:
Hi @Koonweee.

Yes I'm not yet at the end of everything I wanted to test but if all goes well 
I'll be on the weekend from this Thursday evening and I'll be able to finish my tests. 

However, I still have the first results which are rather very encouraging. It's not perfect, 
but it's definitely better. 

However, I ask myself two questions about the small lack to achieve perfection :

- The problem could come from my bed. Clearly on this machine the bed is not great.
A deviation of about 0.4mm (bed from a fysetc kit) and clearly bell-shaped.

- The calibration procedure can on this type of machine (Voron 2.4) induce an error.
I will try to explain clearly.

Let's assume that this is my first use of your function.
To be able to launch your function I must at least make a home, a quad_gantry_level (voron 2.4 obliges)
another home to take into account the correction of the QGLet then I launch the calibration of your function.
However, the measurement reading of your calibration function is based on a QGL which is made with a z-offset 
which is itself biased by the twist of the axis.

When I launch a print I start by activating the compensation then home then QGL (this time with the compensation 
activated) then home then probe of the bed_mesh_calibrate.


The problem is that this QGL carried out with the compensation certainly gives a better mechanical alignment of 
the 4 Z axes but this change of alignment of the Z axes can still modify the angle of the axis and therefore partially 
invalidates the values ​​of compensation.

We could perhaps improve the linearity of the correction by playing on the calculation of the "interpolate_t" variable, 
but frankly this already gives very good results.

Congratulations and thank you again for your work.

fauks commented:
@koonweee I finally got around to redoing it with plastic feeler gauges and it worked perfectly. I can print full plates of parts with no deviation from side to side. 

I think I might have a slight twist in my Y axis as well though, so if we could get this to do both that would be amazing! Deviation from front to back is pretty minor, but noticeable when printing a full plate. Probably less than 0.1mm.

koonweee commented:
> @koonweee I finally got around to redoing it with plastic feeler gauges and it worked perfectly. I can print full plates of parts with no deviation from side to side.
> 
> I think I might have a slight twist in my Y axis as well though, so if we could get this to do both that would be amazing! Deviation from front to back is pretty minor, but noticeable when printing a full plate. Probably less than 0.1mm.

Thanks for the update! Doing it for the Y axis should be possible as well with some refactoring of the code. I am hoping to do this soon (swamped with school at the moment 😓 )

fauks commented:
> > @koonweee I finally got around to redoing it with plastic feeler gauges and it worked perfectly. I can print full plates of parts with no deviation from side to side.
> > I think I might have a slight twist in my Y axis as well though, so if we could get this to do both that would be amazing! Deviation from front to back is pretty minor, but noticeable when printing a full plate. Probably less than 0.1mm.
> 
> Thanks for the update! Doing it for the Y axis should be possible as well with some refactoring of the code. I am hoping to do this soon (swamped with school at the moment 😓 )

That would be awesome! No rush, I appreciate what you have done already!

mhaas commented on klippy/extras/x_twist_compensation.py:
```diff
@@ -0,0 +1,633 @@
+        self.pmgr.create_profile(profile_name, self.results, avg)
+        # recommend z offset to user
+        self.gcmd.respond_info(
+            "X_TWIST_CALIBRATE: Calibration complete, reccomended z_offset: %f"
```
```suggestion
            "X_TWIST_CALIBRATE: Calibration complete, recommended z_offset: %f"
```

mhaas commented on docs/G-Codes.md:
```diff
@@ -1302,6 +1302,43 @@ print.
+is enabled.
+
+#### X_TWIST_CALIBRATE
+`X_TWIST_CALIBRATE [N_POINTS=<value>]`: Initiates the X twist calibration
```
Include `PROFILE_NAME` here?

mhaas commented on klippy/extras/x_twist_compensation.py:
```diff
@@ -0,0 +1,633 @@
+        self.load_profile(profile_name)
+
+    cmd_X_TWIST_PROFILE_CLEAR_help = \
+        "Clears the active mesh"
```
Is mesh the correct wording here?
```suggestion
        "Clears the active X twist compensation"
```

mhaas commented on klippy/extras/x_twist_compensation.py:
```diff
@@ -0,0 +1,633 @@
+        self.clear_profile()
+
+    cmd_X_TWIST_PROFILE_SAVE_help = \
+        "Saves the active mesh to the config file"
```
```suggestion
        "Saves the active X twist compensation profile to the config file"
```

mhaas reviewed (commented):
Drive-by review - currently testing the PR and figured I would check the code as well :)

murdock62 commented:
> 

After more tests I can say that according to my results this function brings a significant improvement.

On a machine where the X axis is fixed like a cr-10 or other, we are almost at perfection.

On a machine like a voron 2.4 there is still a very small defect but I think that as mentioned above 
the calibration being done after a QGL which does not use twist compensation if the QGLs made during 
the prints use this compensation the QGL will be better but the offset difference between the nozzle 
and the probe will no longer be the same and therefore the compensation will not be perfect. 

I think best when printing and only activating the correction after the QGL. 
This QGL will be less precise but will relace the machine with the same twist as during calibration. 
The compensation of the mesh will therefore be better.

Good work !!!

Jolliness0995 commented:
I have a SV06 and I want to let you know that this module has made a huge difference! It seems to be working great for me!

KevinOConnor commented:
Interesting, thanks.  As high-level feedback, compensating for probe location based inaccuracies seems like useful functionality.  When reviewing the code it seems to me like the feature is in an "experimental" phase though.  So, I'm not sure it is ready to merge to the master Klipper repository.

Some high-level comments:
1. It is preferable to add only a few commands with each new module.  I'd expect this module to add an `X_TWIST_COMPENSATE_CALIBRATE` command.  Most of the other commands don't immediately seem necessary to me.
2. It's not clear to me why there is a "profile" system.  Why would a user define multiple "twist profiles".  Seems it would be simpler to have a single twist compensation.
3. It seems odd that this module only applies to X.  I'd imagine some printers would have hardware where the Y axis would be more likely to twist.  (I'm not suggesting the module needs to have more capabilities, but from a module naming perspective it would seem unfortunate to end up in a position where we have multiple "twist compensation" modules just due to axis naming.)

Cheers,
-Kevin

blastrock commented:
Thanks a lot for porting this feature to Klipper! I also have a SV06 and was thinking I had to give up the automatic bed mesh feature when switching from Marlin to Klipper.

I have tested your code (rebased on v0.11.0, but it shouldn't matter much). I have found a couple bugs, you can look at my fixes here https://github.com/blastrock/klipper/commit/0fc4d9ff77bea95f0223b6aee7d30fb40092c238 .

The first bug is that all offsets were calculated against the same measured Z from the probe, at the launch of Klipper. This not only included the X twist offset, but also the offset from the bed not being leveled. It looks like your patch discards the measured Z on each point.

The second bug is that the results array was not reset upon rerunning the calibration. This made the array contain 6 values at the end of the calibration instead of 3.

Really looking forward to this being merged :)

Arksine commented:
Having looked at this PR I agree with Kevin's comments.  I don't think its a good idea to transform saved mesh profiles, instead we should encourage users recalibrate the mesh after twist calibration. 

It may be worth exploring whether or not twist is impacted by heat.  If there is data that shows a measurable impact based on temperature then profiles may be useful.  Should it turn out that profiles are useful I think it would be desirable for the commands to be consistent with `bed_mesh`, where there is a single gcode command responsible for profile management.

blastrock commented:
Hi there,

Since there isn't much activity on this PR, I took the liberty of fixing what was pointed out during this review: https://github.com/Klipper3d/klipper/pull/6149

Flol0 commented:
Is there still someone watching?

Sineos commented:
@Flol0 
It might have made sense to read at least the second last post before posting ...
